Refactors sheet crafting to better support directional construction

vinylspiders opened the pull request:
## About The Pull Request

https://github.com/tgstation/tgstation/blob/0426f7ddbaa91439c7278189101f5db9c7f2ed95/code/game/objects/items/stacks/stack.dm#L449

Ok, but can we not?

This PR refactors sheet crafting to generalize all the cases that were previously locked behind grille/window type checks and such. In their stead there are bitflags that can be set to achieve certain behaviors.

All the behavior from before should be preserved, but now it can be extended to other items. E.g. if you want a railing that can be crafted underneath directional windows, or an item that behaves like a grille does--it's just a matter of setting the right obj_flags for it now. 

This makes it very simple and painless to add new recipes that use directional crafting! It's all modular now.

<details><summary>Details</summary>

---

### What I've done:

-Eliminated all the type checks, instead it will now be handled by object flags and recipe vars, making for a much more configurable system.

-Added two new obj_flags: `BLOCKS_CONSTRUCTION_DIR` and `IGNORE_DENSITY`.
-Additionally, I renamed the existing flag `NO_BUILD` to `BLOCKS_CONSTRUCTION`.

-Changes the proc `valid_window_location` to `valid_build_direction`, and makes it work for things other than windows.

-Removed a deprecated `window_checks` var from the stack_recipe datum.
-Added three more vars to the stack_recipe datum: `check_direction` and `check_density`, `is_fulltile`

-Decoupled `on_solid_ground` from the object density check. Now you can set those separately, allowing you to make recipes that forbid/allow building things over other things while in space.

---

### What the new flags do:

`BLOCKS_CONSTRUCTION` works as before---prevents objects from being built on the object. I felt that the previous name was not descriptive enough, you should know exactly what it does just from looking at the name.
_example: dna scanner_

`BLOCKS_CONSTRUCTION_DIR` -- setting this on an object will prevent objects from being built on it when their directions are the same.
_example: directional windows, windoors, railings_

`IGNORE_DENSITY` -- setting this on an object will cause its density to be ignored when performing the construction density check. This could have other potential uses as well in the future.
_example: grilles, directional windows, tables_

These three flags cover all the bases for the types of items that are currently craftable, so there is no more need for any type checking or weird snowflake window checks. Simply set the appropriate flag and it'll work as you would expect.

---


### What the recipe vars do:
`check_direction` tells the recipe to check if there's something in that direction with the `BLOCKS_CONSTRUCTION_DIR` flag set.

`check_density` tells the recipe to run the density check when set. This is true by default. There are very few items in the game that currently have this set to false--namely grilles. Setting this to false will make it so that the object can be constructed regardless of what is in that tile (unless `one_per_turf` is also set, which will make it so that you can't craft the same thing twice in the same turf).

`is_fulltile` is used for fulltile windows, but it doesn't necessarily have to be--you can give this to any recipe and it will adopt the same properties as that of the fulltile window. Basically they have a special case where they shouldn't be able to be built over directional constructions, where normally things would be able to be. Setting this makes check_direction true as well.

---

### In summary: 

Sheet crafting still works just as it did before. But the backend of it has gotten a glow up and will be able to more easily support new behaviors.

</details>

## Why It's Good For The Game

This makes the crafting system much more flexible to add recipes to, and will prevent bad code practices of stacking more conditionals down the line whenever someone wants to add an item that behaves like grilles or directional windows in how they are constructed.

It had to be done. Those window checks were a mess.

## Changelog

:cl:
qol: added fifty stack versions of remaining glass sheet stacks for ease of debugging
refactor: refactored sheet crafting to better support directional constructions that aren't windows
/:cl:

vinylspiders committed:
Phase 1 - redesigning the logic of building to remove the special cases for windows

To accomplish this: 2 new bitflags
'NO_BUILD_ON_DIRECTION'
'BUILD_ON_IGNORES_DENSITY'

changed 'NO_BUILD' to 'NO_BUILD_ON'

vinylspiders committed:
phase 2 - renaming NO_BUILD in the rest of the files

vinylspiders committed:
phase 3 - degunking the crafting recipe datums

removes deprecated 'window_checks' var

adds some missing 50 stacks for debugging QOL

vinylspiders committed:
Removes the old proc for good

vinylspiders committed:
Final pass, fixes flags for fulltile window subtypes

renames some things to be more descriptive, fixes edge case where fulltile windows need to be SPECIAL.

vinylspiders committed:
I wanted to avoid this but you've left me no choice, fulltile window.

vinylspiders committed:
this will just have to do

vinylspiders committed:
Merge branch 'master' into fix-directional-crafts

vinylspiders committed:
whoop

vinylspiders committed:
Merge branch 'fix-directional-crafts' of https://github.com/vinylspiders/tgstation into fix-directional-crafts

vinylspiders committed:
oops, typo

vinylspiders committed:
some more things that I missed, to make everything behave exactly as before

I'm not sure if the tram wall girders being able to be constructed over other objects is intentional but it's going back to how it was anyway.

vinylspiders committed:
grrr

vinylspiders committed:
Makes fulltile set recipes always run the direction check

This will tidy up the args list a bit.

vinylspiders committed:
Merge branch 'master' into fix-directional-crafts

vinylspiders committed:
damnit stahp adding more instances of this proc I changed the name of

vinylspiders committed:
Still wasn't happy with the bitfield names, so I renamed them again

I think this is better

san7890 commented on code/__DEFINES/obj_flags.dm:
```diff
@@ -11,7 +11,9 @@
 #define BLOCK_Z_IN_DOWN (1<<8) // Should this object block z falling from above?
 #define BLOCK_Z_IN_UP (1<<9) // Should this object block z uprise from below?
-#define NO_BUILD (1<<10) // Can we build on this object?
+#define BLOCKS_CONSTRUCTION (1<<10) // Does this object prevent things from being built on it?
+#define BLOCKS_CONSTRUCTION_DIR (1<<11) // Does this object prevent same-direction things from being built on it?
+#define IGNORE_DENSITY (1<<12) // Can we ignore density when building on this object? (for example, directional windows and grilles)
```
```suggestion
#define BLOCKS_CONSTRUCTION (1<<10) //! Does this object prevent things from being built on it?
#define BLOCKS_CONSTRUCTION_DIR (1<<11) //! Does this object prevent same-direction things from being built on it?
#define IGNORE_DENSITY (1<<12) //! Can we ignore density when building on this object? (for example, directional windows and grilles)
```

secret DMdoc jutsu, we don't really use it but it's great for ensuring that we get these defines docced _somehow_. you could also rewrite all of the flags to use the `///` but this is just in case you don't want to go through that for whatever reason

san7890 commented on code/game/objects/items/stacks/stack_recipe.dm:
```diff
@@ -17,10 +17,17 @@
 	/// If only one of the resulting atom is allowed per turf
 	var/one_per_turf = FALSE
+	/// If the atom is fulltile, as in a fulltile window. This is used for the direction check to prevent fulltile windows from being able to be built over directional stuff.
+	/// Setting this to true will also set check_direction to true.
+	var/is_fulltile = FALSE
```
should say something like "will effectively set X var to true". ideal world is that separate vars should stay separate, but i understand if the construction backend is too complicated to really allow something like this to not happen in some way.

vinylspiders committed:
Update code/__DEFINES/obj_flags.dm

Co-authored-by: san7890 <the@san7890.com>

vinylspiders committed:
comment suggestion

vinylspiders committed:
One very last minute thing that came to mind:

I just realized a lot of recipes (like tiles and such) previously had on_solid_ground set to false, which meant they were able to be crafted while standing in a tile that has an object with density in it.

I'm not sure how often someone is going to need to craft tiles while inside a wall or something but if the functionality is to be preserved these recipes must be adjusted.

Future thought: it would be nice to further refactor the stack_recipe datums into subtypes (like /item, /structure, etc) so the arg lists can be shorter after taking out the ones they generally have in common. But that is too much for one PR.

vinylspiders committed:
Merge branch 'master' into fix-directional-crafts

san7890 commented on code/game/objects/items/stacks/sheets/mineral.dm:
```diff
@@ -313,7 +313,7 @@ GLOBAL_LIST_INIT(titanium_recipes, list ( \
 	walltype = /turf/closed/wall/mineral/plastitanium
 
 GLOBAL_LIST_INIT(plastitanium_recipes, list ( \
-	new/datum/stack_recipe("plastitanium tile", /obj/item/stack/tile/mineral/plastitanium, 1, 4, 20, category = CAT_TILES), \
+	new/datum/stack_recipe("plastitanium tile", /obj/item/stack/tile/mineral/plastitanium, 1, 4, 20, check_density = FALSE, category = CAT_TILES), \
 	))
```
fwiw these backslashes are completely redundant since DM allows multilining lists even if they are in a macro wrapper (metal hydrogen on 466 proves this) you don't have to change anything but i wanted to mention it here because of how common this pattern is
